DuckDuckGo/JSAlertController.swift

```diff
@@ -0,0 +1,208 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import UIKit
+import Core
+
+private extension UIImage {
+    static let highlightedAlertButtonTint = UIImage(named: "AlertButtonHighlightedTint")
+}
+
+final class JSAlertController: UIViewController {
+
+    @IBOutlet var alertView: UIView!
+    @IBOutlet var scrollView: UIScrollView!
+    @IBOutlet var keyboardConstraint: NSLayoutConstraint!
+    @IBOutlet var backgroundView: UIView!
+    @IBOutlet var titleLabel: UILabel!
+    @IBOutlet var messageLabel: UILabel!
+    @IBOutlet var okButton: UIButton!
+    @IBOutlet var cancelButton: UIButton!
+    @IBOutlet var cancelButtonSeparator: UIView!
+    @IBOutlet var closeTabButton: UIButton!
+    @IBOutlet var textField: UITextField!
+    @IBOutlet var textFieldBox: UIView!
+
+    private var alert: WebJSAlert? {
+        didSet {
+            reloadData()
+        }
+    }
+
+    var isShown: Bool {
+        alert != nil
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        reloadData()
+
+        self.alertView.alpha = 0.0
+        self.backgroundView.alpha = 0.0
+
+        self.keyboardConstraint.constant = 0
+        self.registerForKeyboardNotifications()
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        // hide container view
+        if alert == nil {
+            self.view.superview?.isHidden = true
+        }
+    }
+
+    func present(_ alert: WebJSAlert) {
+        self.alert = alert
+
+        self.view.superview?.isHidden = false
+        self.alertView.alpha = 0.0
+        self.alertView.transform = .init(scaleX: 1.15, y: 1.15)
+        self.backgroundView.alpha = 0.0
+
+        UIView.animate(withDuration: 0.2, delay: 0, options: .curveEaseOut) {
+            self.alertView.alpha = 1.0
+            self.alertView.transform = .identity
+            self.backgroundView.alpha = 1.0
+        } completion: { _ in
+
+            if !self.textFieldBox.isHidden {
+                self.textField.becomeFirstResponder()
+            }
+            Pixel.fire(pixel: .jsAlertShown)
+        }
+    }
+
+    override func dismiss(animated: Bool, completion: (() -> Void)? = nil) {
+        if self.textField.isFirstResponder {
+            self.textField.resignFirstResponder()
+        }
+        guard animated else {
+            self.alert = nil
+            self.view.superview?.isHidden = true
+            completion?()
+            return
+        }
+
+        UIView.animate(withDuration: 0.3) {
```

Comment:
The present animation duration is 0.2 but the dismiss is 0.3 is this intentional?

Reply:
yes, this was intentional to match a system alert